openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,615 @@
+    """
+    Create column then in module need them like l10n_de and sale_stock will fill value,
+    https://github.com/odoo/odoo/pull/116643
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_move
+        ADD COLUMN IF NOT EXISTS delivery_date DATE
+        """,
+    )
+
+
+def _am_create_incoterm_location_column(env):
+    """
+    Create column then in sale_stock, purchase_stock will fill it in pre,
+    pr: https://github.com/odoo/odoo/pull/118954
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_move
+        ADD COLUMN IF NOT EXISTS incoterm_location CHARACTER VARYING
+        """,
+    )
+
+
+def _aml_update_invoice_date_like_amount_move(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_move_line
+        ADD COLUMN IF NOT EXISTS invoice_date DATE
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move_line aml
+        SET invoice_date = am.invoice_date
+        FROM account_move am
+        WHERE aml.move_id = am.id
+        """,
+    )
+
+
+def _account_payment_term_migration(env):
+    """
+    https://github.com/odoo/odoo/pull/110274
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_payment_term
+            ADD COLUMN IF NOT EXISTS discount_days INTEGER,
+            ADD COLUMN IF NOT EXISTS discount_percentage FLOAT,
+            ADD COLUMN IF NOT EXISTS early_discount BOOLEAN,
+            ADD COLUMN IF NOT EXISTS early_pay_discount_computation VARCHAR;
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_payment_term_line
+            ADD COLUMN IF NOT EXISTS delay_type VARCHAR,
+            ADD COLUMN IF NOT EXISTS nb_days INTEGER;
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_payment_term_line
+        SET value = 'percent',
+            value_amount = 100.0
+        WHERE value = 'balance'
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_payment_term_line
+        SET delay_type = CASE
+                WHEN end_month = true AND months = 1 THEN 'days_after_end_of_next_month'
+                WHEN end_month = true AND months > 1 THEN 'days_end_of_month_on_the'
+                WHEN end_month IN (false, NULL) THEN 'days_after'
+                ELSE 'days_after'
+            END,
+            nb_days = CASE
+                WHEN months IS NOT NULL AND days_after IS NOT NULL AND end_month = true
+                AND months > 1
+                THEN (months*30) + days + days_after
+                WHEN months IS NOT NULL AND days_after IS NOT NULL AND end_month = true
+                AND months = 1
+                THEN  days + days_after
+                WHEN end_month IN (false, NULL) THEN (months*30) + days
```

Comment:
we can't map all possible configurations of v16 here, so cover the two cases that match exactly, do a heuristic that might be what the user wants for the rest

```suggestion
                WHEN end_month = true AND months <= 1
                THEN COALESCE(days, 0) + COALESCE(days_after, 0)
                ELSE COALESCE(months, 0)*30 + COALESCE(days, 0) + COALESCE(days_after, 0)
```